dali/kernels/common/block_setup.h

```diff
@@ -0,0 +1,263 @@
+    blocks_.clear();
+    is_uniform_ = !force_variable_size && is_uniform(output_shape);
+
+    if (is_uniform_)
+      UniformSizeSetup(output_shape);
+    else
+      VariableSizeSetup(output_shape);
+  }
+
+  TensorListShape<tensor_ndim> GetOutputShape(
+      const TensorListShape<tensor_ndim> &in_shape,
+      span<const TensorShape<ndim>> output_sizes) {
+    TensorListShape<tensor_ndim> shape;
+    shape.resize(in_shape.num_samples(), tensor_ndim);
+    for (int i = 0; i < in_shape.num_samples(); i++) {
+      auto out_tshape = shape.tensor_shape_span(i);
+      int in_d = 0;
+      for (int j = 0; j < tensor_ndim; j++) {
+        out_tshape[j] = (j == channel_dim)
+         ? in_shape.tensor_shape_span(i)[channel_dim]
+         : output_sizes[i][in_d++];
+      }
+    }
+    return shape;
+  }
+
+
+  void ValidateOutputShape(
+      const TensorListShape<tensor_ndim> &out_shape,
+      const TensorListShape<tensor_ndim> &in_shape,
+      span<const TensorShape<ndim>> output_sizes) {
+    TensorListShape<tensor_ndim> shape;
+    shape.resize(in_shape.num_samples(), tensor_ndim);
+    for (int i = 0; i < in_shape.num_samples(); i++) {
+      auto out_tshape = out_shape[i];
+      TensorShape<tensor_ndim> expected_shape;
+
+      int in_d = 0;
+      for (int j = 0; j < tensor_ndim; j++) {
+        expected_shape[j] = (j == channel_dim)
+         ? in_shape.tensor_shape_span(i)[channel_dim]
+         : output_sizes[i][in_d++];
+      }
+
+      DALI_ENFORCE(out_tshape == expected_shape,
+        "Invalid output tensor shape for sample: " + std::to_string(i));
+    }
+  }
+
+  dim3 BlockDim() const {
+    return dim3(block_dim_.x, block_dim_.y, block_dim_.z);
+  }
+
+  ivec3 BlockDimVec() const {
+    return block_dim_;
+  }
+
+  void SetBlockDim(ivec3 block_dim) {
+    block_dim_ = block_dim;
+  }
+
+  void SetBlockDim(dim3 block_dim) {
+    block_dim_ = { block_dim.x, block_dim.y, block_dim.z };
+  }
+
+  dim3 GridDim() const {
+    return dim3(grid_dim_.x, grid_dim_.y, grid_dim_.z);
+  }
+
+  ivec3 GridDimVec() const {
+    return grid_dim_;
+  }
+
+  void SetDefaultBlockSize(ivec<ndim> block_size) {
+    default_block_size_ = block_size;
+    max_block_elements_ = volume(block_size);
+  }
+
+  span<const BlockDesc> Blocks() const { return make_span(blocks_); }
+
+  ivec<ndim> UniformOutputSize() const {
+    assert(is_uniform_);
+    return uniform_output_size_;
+  }
+
+  ivec<ndim> UniformBlockSize() const {
+    assert(is_uniform_);
+    return uniform_block_size_;
+  }
+
+  bool IsUniformSize() const { return is_uniform_; }
+
+  static inline ivec<ndim> shape2size(const TensorShape<tensor_ndim> &shape) {
+    return shape2vec(skip_dim<channel_dim>(shape));
+  }
+
+ private:
+  std::vector<BlockDesc> blocks_;
+  ivec3 block_dim_{32, 32, 1};
+  ivec3 grid_dim_{1, 1, 1};
+  ivec<ndim> uniform_block_size_, uniform_output_size_;
+  ivec<ndim> default_block_size_;
+  int max_block_elements_ = 1<<16;
```

Comment:
Wouldn't it be better to set this in the constructor as `volume(default_block_size_)`?

Reply:
It would.